### PR TITLE
ARROW-12063: [C++] Add null placement option to sort functions

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@ianmcook @jonkeane You may want to review the docs and wordings.

**Comment by pitrou** [comment]
@ursabot please benchmark

**Review by lidavidm — APPROVED** [review]
LGTM, thanks for pushing through all this.

**Comment by ianmcook** [comment]
LGTM, thank you!

I'm curious why use a `NullPlacement` enum to instead of just a `nulls_last` Boolean. Might the enum be extended to include other options in the future (for example controlling whether NaNs come before or after nulls)?

**Comment by lidavidm** [comment]
Even if we don't plan on extending the enum, personally I find something like `NullPlacement::NullsFirst` more readable than `/*nulls_last=*/false`. Especially since it's consistent with `SortOrder::Ascending/Descending` which you could otherwise argue is a boolean (but is more readable as an enum).

**Comment by ianmcook** [comment]
Got it, makes sense, thanks.

File: docs/source/cpp/compute.rst [hunk 1]
```diff
@@ -1475,16 +1480,13 @@ value, but smaller than nulls.
 * \(2) The output is an array of indices into the input, that define a
   stable sort of the input.
 
-* \(3) Input values are ordered lexicographically as bytestrings (even
-  for String arrays).
-
-* \(4) The input must be an array. The default order is ascending.
+* \(3) The input must be an array. The default order is ascending.
```
Comment (ianmcook) [inline docs/source/cpp/compute.rst:1483]: Consider adding here: "By default, nulls and NaNs are sorted after any non-null values."

Comment (pitrou) [reply docs/source/cpp/compute.rst:1483]: It's already above in the section summary.

**Pull request closed without merging**